File: Stack/155. Min Stack/Solution.java
import java.util.*;
import java.lang.Math;

// N : no of elements in a stack
// Time Complexity : Creating MinStack Object - O(1), push() - O(1), pop() - O(1), top() - O(1), getMin() - O(N)
// Space Complexity : O(N)
// class MinStack {
//     Stack<Integer> stack;

//     public MinStack() {
//         stack = new Stack<>();
//     }
    
//     public void push(int val) {
//         stack.push(val);
//     }
    
//     public void pop() {
//         stack.pop();
//     }
    
//     public int top() {
//         return stack.peek();
//     }
    
//     public int getMin() {
//         return stack.stream().min((a, b)->a.compareTo(b)).orElse(Integer.MAX_VALUE);
//     }
// }



// N : no of elements in a stack
// Time Complexity : Creating MinStack Object - O(1), push() - O(1), pop() - O(1), top() - O(1), getMin() - O(1)
// Space Complexity : O(N)
class MinStack {
    Stack<Integer> mainStack;
    Stack<Integer> minStack;

    public MinStack() {
        mainStack = new Stack<>();
        minStack = new Stack<>();
    }
    
    public void push(int val) {
        mainStack.push(val);
        if (minStack.isEmpty()) minStack.push(val);
        else minStack.push(Math.min(val, minStack.peek()));
    }
    
    public void pop() {
        mainStack.pop();
        minStack.pop();
    }
    
    public int top() {
        return mainStack.peek();
    }
    
    public int getMin() {
        return minStack.peek();
    }
}

File: Stack/155. Min Stack/Solution.py
# N : no of elements in a stack
# Time Complexity : Creating MinStack Object - O(1), push() - O(1), pop() - O(1), top() - O(1), getMin() - O(N)
# Space Complexity : O(N)
# class MinStack:

#     def __init__(self):
#         self.stack = []

#     def push(self, val: int) -> None:
#         self.stack.append(val)

#     def pop(self) -> None:
#         self.stack.pop()

#     def top(self) -> int:
#         return self.stack[-1]

#     def getMin(self) -> int:
#         return min(self.stack)


# N : no of elements in a stack
# Time Complexity : Creating MinStack Object - O(1), push() - O(1), pop() - O(1), top() - O(1), getMin() - O(N)
# Space Complexity : O(N)
class MinStack:

    def __init__(self):
        self.mainStack = []
        self.minStack = []

    def push(self, val: int) -> None:
        self.mainStack.append(val)
        if len(self.minStack) == 0:
            self.minStack.append(val)
        else:
            self.minStack.append(min(val, self.minStack[-1]))

    def pop(self) -> None:
        self.mainStack.pop()
        self.minStack.pop()

    def top(self) -> int:
        return self.mainStack[-1]

    def getMin(self) -> int:
        return self.minStack[-1]